File: csharp/Kelson.Advent/Day1/RocketEquation.cs
using System.Collections.Generic;
using System.Linq;

namespace Kelson.Advent.Day1
{
    public static class RocketEquation
    {
        public static int FuelRequired(int mass) => mass / 3 - 2;

        public static int ModuleFuelRequirement(this IEnumerable<int> moduleMasses) => moduleMasses.Select(FuelRequired).Sum();

        public static int TotalFuelRequired(this int mass)
        {
            int fuel_required = FuelRequired(mass);
            if (fuel_required <= 0)
                return 0;
            else
                return fuel_required + TotalFuelRequired(fuel_required);
        }

        public static int TotalModuleFuelRequirement(this IEnumerable<int> moduleMasses) => moduleMasses.Select(TotalFuelRequired).Sum();
    }
}

File: kotlin/src/kelson/advent/day1/RocketEquation.kt
package kelson.advent.day1

import kelson.advent.runAsTest
import java.io.File
import kotlin.test.assertEquals

fun fuelRequired(mass : Int) : Int = mass / 3 - 2

fun moduleFuelRequirement(modules : Iterable<Int>) = modules.map{ fuelRequired(it) }.sum()

fun totalFuelRequired(mass : Int) : Int {
    val fuelRequired = fuelRequired(mass)
    return if (fuelRequired <= 0)
        0
    else
        fuelRequired + totalFuelRequired(fuelRequired)
}

fun totalModuleFuelRequirement(modules : Iterable<Int>) = modules.map{ totalFuelRequired(it) }.sum()

fun testPart1ForModuleFuelRequirements() =
    listOf(
        Pair(12, 2),
        Pair(14, 2),
        Pair(1969, 654),
        Pair(100756, 33583))
    .map{ (mass, fuel) -> Pair(fuelRequired(mass), fuel) }
    .forEach { (required, expected) -> assertEquals(required, expected) }

fun testPart2ForRecursiveFuelRequirements() =
    listOf(
        Pair(14, 2),
        Pair(1969, 966),
        Pair(100756, 50346))
    .map{ (mass, fuel) -> Pair(totalFuelRequired(mass), fuel) }
    .forEach { (required, expected) -> assertEquals(required, expected) }


fun runTests() {
    println("Day 1:")
    runAsTest("Part 1", ::testPart1ForModuleFuelRequirements)
    runAsTest("Part 2", ::testPart2ForRecursiveFuelRequirements)
}

fun run() {
    println("--- DDay 1: The Tyranny of the Rocket Equation ---")
    val modules = File("../resources/day1/input.txt").readLines().map(String::toInt)

    println("Part 1 - Required fuel for modules:")
    val total = moduleFuelRequirement(modules)
    println(total)

    println("Part 2 - Required fuel for modules and fuel:")
    val biggerTotal = totalModuleFuelRequirement(modules)
    println(biggerTotal)
}